File: Python/98_ValidateBinarySearchTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    
    def isValidBST(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: bool
        """

        def dfs(root):
            if root == None:# base
                return True
            else:
                res = True
                if root.left != None:
                    if root.left.val < root.val:
                        res = dfs(root.left)
                    else:
                        return False

                if root.right != None:
                    if root.val < root.right.val:
                        res = dfs(root.right)
                    else:
                        return False
            
            return res

        return dfs(root)
        
# algorithm DFS
# condition: light < root < right
# time O(N) N is number of nodes
# memory O(N) 

File: C++/98_ValidateBinarySearchTree.cpp
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isValidBSTwithMinMax(TreeNode* root, long long min, long long max){
        if (!root) return true;
        else{
            if (min >= root->val || root->val >= max) return false; 

            return (isValidBSTwithMinMax(root->left, min, root->val) 
            && isValidBSTwithMinMax(root->right, root->val, max));
        }
    }

    bool isValidBST(TreeNode* root) {
        return isValidBSTwithMinMax(root, LLONG_MIN, LLONG_MAX);
    }

};